Marlin/Marlin_main.cpp

```diff
@@ -2419,8 +2425,13 @@ static void clean_up_after_endstop_or_probe_move() {
 
     #elif ENABLED(AUTO_BED_LEVELING_UBL)
 
+      #if ENABLED(UBL_DELTA)
+        if (( ubl.state.active ) && ( ! enable )) {   // leveling from on to off
+          planner.unapply_leveling(current_position);
```

Comment:
@thinkyhead, even after studying the code, I'm really confused about the intention and proper maintenance of current_position with bed leveling on, off, or transitioning.  I'd like to understand this so I can make sure I get it right with respect to UBL changes.

The code (with your fix) looks like this:
```
#if HAS_LEVELING
  /**
   * Turn bed leveling on or off, fixing the current
   * position as-needed.
   *
   * Disable: Current position = physical position
   *  Enable: Current position = "unleveled" physical position
   */
  void set_bed_leveling_enabled(bool enable/*=true*/) {
    #elif ENABLED(AUTO_BED_LEVELING_UBL)
      #if PLANNER_LEVELING
        if (ubl.state.active != enable) {
          if (!enable)   // leveling from on to off
            planner.apply_leveling(current_position[X_AXIS], current_position[Y_AXIS], current_position[Z_AXIS]);
          else
            planner.unapply_leveling(current_position);
        }
      #endif
      ubl.state.active = enable;

```
So does current_position[Z] always reflect the logical unleveled position of the nozzle with respect to the last move completed?  It would make sense that if someone did G1 Z1 then asked where the nozzle was located would reply Z 1.0 regardless of leveling applied.

And, prepare_kinematic_line_to (preceding UBL and where I copied logic from for UBL), starts the first segment at logical current_position[Z], with the final target of a full line is specified in unleveled Z.  For extruding a logically level line then, both the current_position[Z] and target[Z] should be the same, and all per-segment leveling occurs below this "logical" level.  The steppers are adjusted for leveling, but the current_position still holds virtual unleveled Z position.

Now if we turn leveling from off (where steppers and current_position[z] should match) to on (but haven't performed any move yet), then current_position[Z] will still reflect unleveled Z and steppers will also reflect unleveled Z.  Then if we do a move, the first leveling gets applied and steppers will end up with leveled Z while current_position again has unleveled Z.

Now if we turn leveling from on to off, the steppers MAY or MAY NOT have the leveled Z depending on whether they ever moved after leveling was enabled (versus turn leveling on then off without moving).

So at the time leveling is turned from off to on, it would make sense to actually MOVE the steppers to the z-leveled current_position, so that current_position[Z] is consistent with stepper-leveled Z.  And likewise, if leveling is turned from on to off, a Z move back to unleveled Z still at current_position would make sense to keep everything consistent.

But it looks like instead that set_bed_leveling_enabled is manipulating current_position[Z] using apply_leveling or unapply_leveling, but I would think current_position[Z] should remain unchanged, still reflecting the logical Z position of last move.

So is current_position supposed to track the unleveled "logical" Z?  Or the physical stepper-represented leveled Z?  It can't really be the leveled-Z because then each kinematic move would start from the wrong Z (it assumes current_position is unleveled, same as destination).

Sorry I'm a bit slow on the uptake on this area and I would like to make sure I understand it correctly.  Thanks for any insight you can provide.